docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+
+* Written size -- that is basically `last_received_lsn`,
+* Synthetic storage size -- there is a way to calculate it, albeit a costly one,
+* Real storage size -- there is a way to calculate it using a layer map or filesystem,
+* S3 storage size -- can calculate it by S3 API calls
+
+Some of those metrics are expensive to calculate, so the reporting period here is driven mainly by implementation details. We can set it to, for example, once per hour. Not a big deal since the pageserver is stateful, and all metrics can be reported as an absolute value, not increments. At the same time, a smaller reporting period improves UX, so it would be good to have something more real-time.
+
+`written size` is primarily a safekeeper-related metric, but since it is available on both pageserver and safekeeper, we can avoid reporting anything from the safekeeper.
+
+Example event:
+
+```json
+{
+    "metric": "s3_tenant_size_bytes",
+    "type": "absolute",
+    "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+    "time": ...,
+    "value": 12345454,
+}
+```
+
+# Data collection
+
+## Push vs. pull
+
+We already have pull-based Prometheus metrics, so it is tempting to use them here too. However, in our setup, it is hard to tell when some metric changes. For example, garbage collection will constantly free some disk space over a week, even if the project is down for that week. We could also iterate through all existing tenants/branches/endpoints, but that means some amount of code to do that properly and most likely we will end up with some per-metric hacks in the collector to cut out some of the tenants that are surely not changing that metric.
+
+With the push model, it is easier to publish data only about actively changing metrics -- pageserver knows when it performs s3 offloads, garbage collection and starts/stops consuming data from the safekeeper; proxy knows about connected clients; console / autoscaler-agent knows about active cpu time.
+
+Hence, let's go with a push-based model.
+
+## Common bus vs. proxying through the console
+
+We can implement such push systems in a few ways:
+
+a. Each component pushes its metrics to the "common bus", namely segment, Kafka, or something similar. That approach scales well, but it would be harder to test it locally, will introduce new dependencies, we will have to distribute secrets for that connection to all of the components, etc. We would also have to loop back some of the events and their aggregates to the console, as we want to show some that metrics to the user in real-time.
+
+b. Each component can call HTTP `POST` with its events to the console, and the console can forward it to the segment for later integration with metronome / orb / onebill / etc. With that approach, only the console has to speak with segment. Also since that data passes through the console, the console can save the latest metrics values, so there is no need for constant feedback of that events back from the segment.
```

Comment:
>  Each component can call HTTP `POST` with its events to the console, and the console can forward it to the segment for later integration with metronome / orb / onebill / etc. With that approach, only the console has to speak with segment. Also since that data passes through the console, the console can save the latest metrics values, so there is no need for constant feedback of that events back from the segment.

this approach makes the console a single point of failure.
it forces to heavily optimize for the lack of any downtimes of the console and brings up hard questions about the error handling strategies.

what should be the strategies for the other components for handling the failures of the POSTs? retry?
i think there could be the situations when the sender thinks that the message wasn't delivered, when in fact it was.

> the console can save the latest metrics values

with the deltas approach, not only the latest metrics have to be sent, but also the data potentially needs to be aggregated with the messages and measurements received previously. 
horizontal scaling of the processing in the console may be interesting, given the questions about the retries and possibilities of multiple deliveries.

Reply:
> this approach makes the console a single point of failure.

It is already. As well as the console database. We can later split that collector to a separate service, but for now I think it fine, given that we write data to the database and segment.

Or we can just report to segment right from the service. I thought about that initially, but then it is tricky for the console to subscribe to that events, and not sure moving data through the segment will improve our reliability in practical terms.

> horizontal scaling of the processing in the console may be interesting

that should just work since we touch database (or redis optionally)

> what should be the strategies for the other components for handling the failures of the POSTs? retry?
i think there could be the situations when the sender thinks that the message wasn't delivered, when in fact it was.

sender should consider data as delivered only when it received confirmation from the collector. That will result in at least once semantics. On failure/timeout we can just avoid resetting delta counter / last send at, so instead of retry we will report delta over a bigger time frame.

Reply:
> > this approach makes the console a single point of failure.
> 
> It is already.

it is for establishing new connections, but now it's on a critical path for accurate billing too..

> Or we can just report to segment right from the service. I thought about that initially, but then it is tricky for the console to subscribe to that events, and not sure moving data through the segment will improve our reliability in practical terms.

could be segment or kafka or something else.


> > i think there could be the situations when the sender thinks that the message wasn't delivered, when in fact it was.
> 
> On failure/timeout we can just avoid resetting delta counter / last send at, so instead of retry we will report delta over a bigger time frame.

we need to perform careful testing with all of our LBs and reverse proxies in place to make sure the sender does think the data wasn't accepted when in fact it was.
going away from deltas could help with the case that I mentioned too.